File: JS/Tema9_Ejercicio4.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 4
 * ========================================================
 * Métodos estáticos y propiedades de clase.
 * 
 * EXPLICACIÓN:
 * - static: método/propiedad de la clase, no de instancia
 * - Se accede con NombreClase.metodo()
 * - No necesita crear instancia
 * - Caso de uso: funciones de utilidad, contadores
 */

class Contador {
    static numInstancias = 0;
    
    constructor(nombre) {
        this.nombre = nombre;
        Contador.numInstancias++;
    }
    
    static obtenerInstancias() {
        return `Total de instancias: ${Contador.numInstancias}`;
    }
    
    static resetear() {
        Contador.numInstancias = 0;
    }
}

class Matematica {
    static PI = 3.14159;
    
    static cuadrado(n) {
        return n * n;
    }
    
    static cubo(n) {
        return n * n * n;
    }
    
    static areaCirculo(radio) {
        return this.PI * radio * radio;
    }
}

console.log("=== MÉTODOS ESTÁTICOS ===\n");

console.log("Clase Matemática:");
console.log(`5² = ${Matematica.cuadrado(5)}`);
console.log(`3³ = ${Matematica.cubo(3)}`);
console.log(`Área círculo r=4: ${Matematica.areaCirculo(4).toFixed(2)}`);

console.log("\nContador:");
const obj1 = new Contador("A");
const obj2 = new Contador("B");
const obj3 = new Contador("C");

console.log(Contador.obtenerInstancias());

Contador.resetear();
console.log("Después de resetear:");
console.log(Contador.obtenerInstancias());

/*
 * SALIDA ESPERADA:
 * === MÉTODOS ESTÁTICOS ===
 * 
 * Clase Matemática:
 * 5² = 25
 * 3³ = 27
 * Área círculo r=4: 50.27
 * 
 * Contador:
 * Total de instancias: 3
 * Después de resetear:
 * Total de instancias: 0
 */

File: Python/Tema9_Ejercicio4.py
"""
TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS
EJERCICIO 4: MÉTODOS ESTÁTICOS Y DE CLASE
===========================================

Métodos que pertenecen a la clase, no a instancias
"""

# MÉTODOS ESTÁTICOS
print("=== MÉTODOS ESTÁTICOS ===\n")

class Matematica:
    @staticmethod
    def sumar(a, b):
        return a + b
    
    @staticmethod
    def multiplicar(a, b):
        return a * b
    
    @staticmethod
    def es_par(n):
        return n % 2 == 0

print(f"5 + 3 = {Matematica.sumar(5, 3)}")
print(f"5 * 3 = {Matematica.multiplicar(5, 3)}")
print(f"4 es par: {Matematica.es_par(4)}")

# MÉTODOS DE CLASE (CLASSMETHOD)
print("\n=== MÉTODOS DE CLASE ===\n")

class Contador:
    instancias = 0
    
    def __init__(self, nombre):
        self.nombre = nombre
        Contador.instancias += 1
    
    @classmethod
    def obtener_total(cls):
        return cls.instancias

c1 = Contador("A")
c2 = Contador("B")
c3 = Contador("C")

print(f"Total de instancias: {Contador.obtener_total()}")

# COMBINACIÓN
print("\n=== COMBINACIÓN ===\n")

class Configuracion:
    idioma = "español"
    volumen = 80
    
    @classmethod
    def cambiar_idioma(cls, nuevo_idioma):
        cls.idioma = nuevo_idioma
        print(f"Idioma cambiado a: {nuevo_idioma}")
    
    @staticmethod
    def validar_volumen(vol):
        return 0 <= vol <= 100

print(f"Idioma: {Configuracion.idioma}")
Configuracion.cambiar_idioma("inglés")
print(f"Volumen válido: {Configuracion.validar_volumen(50)}")

"""
SALIDA ESPERADA:
=== MÉTODOS ESTÁTICOS ===

5 + 3 = 8
5 * 3 = 15
4 es par: True

=== MÉTODOS DE CLASE ===

Total de instancias: 3

=== COMBINACIÓN ===

Idioma: español
Idioma cambiado a: inglés
Volumen válido: True
"""
